### PR TITLE
[TT-Train] Update tt-train/nanoGPT example to use TTNN ops for sampling

### PR DESCRIPTION
**Pull request by abustamanteTT**
### Problem description
Updates TT-Train NanoGPT example to use TT ops for hardware acceleration

### What's changed
Previous implementation of nanoGPT example used CPU-based sampling. Now composite sampling using TT ops and TT argmax

### Checklist
https://github.com/tenstorrent/tt-metal/actions/runs/17868336709
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 1]
```diff
@@ -280,62 +154,89 @@ void generate(
...
     std::vector<uint32_t> prompt_tokens_padded(max_sequence_length, pad_token_id);
+    std::vector<float> padded_logits_vector(padded_vocab_size, 0.0F);
 
     fmt::print("Generated text:\n");
     fmt::print("*******************\n");
     fmt::print("{}", prompt);
 
+    // Sampling setup
+
+    auto repeats = ttnn::Shape({1U, 1U, SamplingBatchSize, 1U});
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp]: please remove dead code.

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 2]
```diff
@@ -76,6 +76,28 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
+
+    auto rand = ttnn::rand(
+        ttnn::DefaultQueueId,
+        t.logical_shape(),
+        *device,
+        t.dtype(),
+        t.layout(),
+        ttnn::types::DRAM_MEMORY_CONFIG,
+        0.00001F,
+        0.99F,
```
[the remark this answers is not in the document]
Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: please add names of the variable like: /* device*/ device

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 3]
```diff
@@ -76,6 +76,28 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
+        ttml::autograd::ctx().get_generator()());
+
+    rand = ttnn::neg(ttnn::log(ttnn::neg(ttnn::log(rand))));
+    auto out = ttnn::mul_sfpu(t, 1.0F / (temperature > 0.0F ? temperature : 1.0F));
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: Please dont apply my code if temperature is 0.0F or less. Just call argmax.

**Review by dmakoviichuk-tt — COMMENTED** [review]
Please add unit tests for ttnn_fixded sample. Overall good.

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 4]
```diff
@@ -6,6 +6,8 @@
 #include <core/ttnn_all_includes.hpp>
 #include <csignal>
 #include <cstdint>
+#include <numeric>
+#include <tt_stl/small_vector.hpp>
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp]: please check, you might not need it.

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 5]
```diff
@@ -76,6 +76,28 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
     return ttnn::sum(t, dim, keep_dim, std::nullopt, core::ComputeKernelConfig::precise());
 }
 
+tt::tt_metal::Tensor sample(const tt::tt_metal::Tensor& t, const tt::tt_metal::Tensor& argmax_mask, float temperature) {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: probably argmax_mask could be optional in this function.

Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: [f26c081](https://github.com/tenstorrent/tt-metal/pull/28732/commits/f26c0812d2143346a6dbe79191d111382be477a9)

File: tt-train/sources/examples/nano_gpt/main.cpp (hunk 4 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/nano_gpt/main.cpp]: Commit [f26c081](https://github.com/tenstorrent/tt-metal/pull/28732/commits/f26c0812d2143346a6dbe79191d111382be477a9)

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 3 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: Commit [f26c081](https://github.com/tenstorrent/tt-metal/pull/28732/commits/f26c0812d2143346a6dbe79191d111382be477a9)

File: tt-train/sources/examples/nano_gpt/main.cpp (hunk 1 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/nano_gpt/main.cpp]: Commit [f26c081](https://github.com/tenstorrent/tt-metal/pull/28732/commits/f26c0812d2143346a6dbe79191d111382be477a9)

File: tt-train/tests/ttnn_fixed/slice_op_test.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    // Example: slice out the last row, columns 1 to 5
+    uint32_t H_out = H;
+    uint32_t W_out = W;
+    ttnn::SmallVector<uint32_t> step = {1U, 1U, 1U, 1U};
+    ttnn::SmallVector<uint32_t> start_index = {0U, 0U, H - 1, 1U};
+    ttnn::SmallVector<uint32_t> end_index = {N, C, H, 5U};
+
+    // Expected output: shape {1, 1, 1, 4}, values from a(0,0,H-1,1) to a(0,0,H-1,4)
+    xt::xarray<float> b = {{{
+
```
Comment (rfurko-tt) [inline tt-train/tests/ttnn_fixed/slice_op_test.cpp]: seems like you have problem with clang-tidy

File: tt-train/tests/ttnn_fixed/slice_op_test.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <gtest/gtest.h>
+
+#include <core/ttnn_all_includes.hpp>
+#include <iostream>
+#include <ttnn/operations/data_movement/slice/slice.hpp>
+#include <ttnn/operations/reduction/generic/generic_reductions.hpp>
```
Comment (rfurko-tt) [inline tt-train/tests/ttnn_fixed/slice_op_test.cpp]: use ttnn_all_includes header

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: Commit c7df2dc

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 8]
```diff
@@ -76,6 +76,36 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
+            out.logical_shape(),              // tensor shape
+            *device,                          // MeshDevice
+            out.dtype(),                      // tensor datatype
+            out.layout(),                     // tensor layout
+            ttnn::types::DRAM_MEMORY_CONFIG,  // tensor memory config
+            0.00001F,                         // minimum value
+            0.99F,                            // maximum value
+            seed);                            // seed
+
+        rand = ttnn::neg(ttnn::log(ttnn::neg(ttnn::log(rand))));
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp:101]: add short comment about trick you've used here

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 9]
```diff
@@ -76,6 +76,36 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
 
+tt::tt_metal::Tensor sample(
+    const tt::tt_metal::Tensor& t, float temperature, uint32_t seed, std::optional<tt::tt_metal::Tensor> argmax_mask) {
+    auto* device = &ttml::autograd::ctx().get_device();
+
+    ttnn::Tensor out = t;
+
+    if (temperature > 0.0F) {
+        auto rand = ttnn::rand(
+            ttnn::DefaultQueueId,             // queue ID
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: I think you don't need to pass `ttnn::DefaultQueueId`, just remove, no other changes are required

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 10]
```diff
@@ -76,6 +76,36 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
+            out.logical_shape(),              // tensor shape
+            *device,                          // MeshDevice
+            out.dtype(),                      // tensor datatype
+            out.layout(),                     // tensor layout
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: we use this style of argument comments `/* layout */ out.layout()`

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp [hunk 11]
```diff
@@ -76,6 +76,36 @@ tt::tt_metal::Tensor sum_ttnn(const tt::tt_metal::Tensor& t, int dim, bool keep_
...
+            0.99F,                            // maximum value
+            seed);                            // seed
+
+        rand = ttnn::neg(ttnn::log(ttnn::neg(ttnn::log(rand))));
+        out = ttnn::mul_sfpu(out, 1.0F / temperature);
+        out = ttnn::add(out, rand);
+    }
+
+    if (argmax_mask.has_value()) {
+        out = ttnn::subtract(out, argmax_mask.value());
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: why do we subtract argmax_mask? maybe we should find a better name for this variable

File: tt-train/sources/examples/nano_gpt/main.cpp [hunk 12]
```diff
@@ -280,62 +147,88 @@ void generate(
...
+        if (next_token_id >= original_vocab_size) {
+            next_token_id = prompt_tokens.back();
+        }
 
         // Append the new token
         prompt_tokens.push_back(next_token_id);
 
         // Decode and print
         fmt::print("{}", tokenizer.decode({next_token_id}));
+        std::cout.flush();
```
Comment (rfurko-tt) [inline tt-train/sources/examples/nano_gpt/main.cpp:231]: is it required? did we see any issue without flush?

**Review by rfurko-tt — COMMENTED** [review]
Looks good overall, a few things to clean up

File: tt-train/sources/examples/nano_gpt/main.cpp (hunk 12 above)
Comment (abustamanteTT) [reply tt-train/sources/examples/nano_gpt/main.cpp:231]: Not strictly an issue, was added for visibility of speedup from CPU sampling -> device sampling

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 11 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: It's more of an inverse mask; we want to blow all logits associated with padded vocab size away in cases where the original vocab doesn't fit neatly into tiles

Comment (rfurko-tt) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: it's just the suffix `_mask` that I don't like.

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 9 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: Doesn't build without `ttnn::DefaultQueueId`, I don't see other function signatures for `ttnn::rand` that ignores queue ID

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 10 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: Commit [b589550](https://github.com/tenstorrent/tt-metal/pull/28732/commits/b589550c80f823f748a166f50baa4aa05e235d50)

File: tt-train/tests/ttnn_fixed/slice_op_test.cpp (hunk 6 above)
Comment (abustamanteTT) [reply tt-train/tests/ttnn_fixed/slice_op_test.cpp]: Commit [b589550](https://github.com/tenstorrent/tt-metal/pull/28732/commits/b589550c80f823f748a166f50baa4aa05e235d50)

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 8 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp:101]: Commit [b589550](https://github.com/tenstorrent/tt-metal/pull/28732/commits/b589550c80f823f748a166f50baa4aa05e235d50)

File: tt-train/tests/ttnn_fixed/slice_op_test.cpp (hunk 7 above)
Comment (abustamanteTT) [reply tt-train/tests/ttnn_fixed/slice_op_test.cpp]: Commit [070908e](https://github.com/tenstorrent/tt-metal/pull/28732/commits/070908e9061b63ff4d66e19cd1a2d85d90245e9b)

File: tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp (hunk 9 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ttnn_fixed/trivial_ttnn_ops.cpp]: My mistake, build error was from something else. Fixed in commit [09b2eb6](https://github.com/tenstorrent/tt-metal/pull/28732/commits/09b2eb61eb8820f97fb5df177a4917074b8c9894)

**Review by rfurko-tt — APPROVED** [review]
